### PR TITLE
A method to switch off mandatory prime number limit

### PR DESCRIPTION
**Pull request by InfoHunter**
A major use of multi-prime RSA key is to achieve performance promotions and reduce hardware costs. But PR #4791 set a mandatory limit to prevent generating more primes for an RSA key. For instance, it only allows 2048-bit RSA key with 3 primes, which makes no practical usage since very few users want to use a 3-prime-2048 key to replace a 2-prime-2048 key for only 20% - 25% performance promotion. This makes this feature in OpenSSL almost useless for a lot of users as far as I know.

Someone will still argue about 10-prime-2048 is definitely less mathematical-secure than 3-prime-2048 (which almost all people know, and so do I...), but that's not the reason we should **force** the users to use 3-prime-2048, especially the users who require this feature may be very huge... 

So here comes this PR:

A new flag RSA_FLAG_INSECURE_PRIMES is introduced to use to switch off the prime number's limit in multi-prime RSA key generation and check functions. If a user knows the risk, then he/she can just set this simple flag and he/she will be free to generate keys...otherwise lots of people need to modify the code and maintain customized patch...

We should **guide** the users to choose what is the most suitable stuff, this is why I still leave the security limit introduced in #4791 there. Again as I mentioned lots of times, the most secure thing in mathematic may not be more secure than other less secure thing in mathematic in practical circumstances...

Documentaion and test cases are also updated to support this method.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by t-j-h** [comment]
This matches what we discussed on the multi-prime adding code (where it was approved with no limits and then limits were subsequently added omitting a mechanism to disable such a limit). Allowing a user to elect to not have the limits on the number of primes is appropriate IMHO.

File: doc/man1/genrsa.pod [hunk 1]
```diff
@@ -96,6 +97,14 @@ RSA key, which is defined in RFC 8017.
 The size of the private key to generate in bits. This must be the last option
 specified. The default is 2048.
 
+=item B<-insecure>
+
+By default OpenSSL will force users to generate secure multi-prime keys only.
+That is to say, the number of primes is fixed due to a security policy, see
```
Comment (paulidale) [inline doc/man1/genrsa.pod]: Is it fixed or just limited above?

Overall, I'd like to see a bit more text explaining why multi-prime keys are weaker.  We're expecting them to make an informed choice without providing any hint as to the trade offs involved.

All that said, the code looks fine.

**Comment by dot-asm** [comment]
There are only two angles to it, it's either

- inadequate and **indefensible** design choice;
- backdooring;

**Comment by richsalz** [comment]
Or working with deployed systems.

**Comment by dot-asm** [comment]
> Or working with deployed systems.

Which is either poorly designed or backdoored. As already said [in #4241] I fail to see how does it mean that we should do same.

**Comment by richsalz** [comment]
Or legacy.  Our policy is that we are secure by default but let users override things if they need; here, too

**Comment by dot-asm** [comment]
> Or legacy.

What legacy? It wasn't released... And as far as external legacies go, it's all about communicating with legacy systems or managing your data at rest. Can OpenSSL communicate with system that uses excessive amount of primes? Yes(*), even as is. Does cap on amount of primes affect your chances to verify old signatures or generate new ones with old key? No, not in any way. Also a word about legacy system as term. It's more or less implied that it's defined by *consensual* recognition that it was *good* idea at the time, i.e. we didn't know better. Is it actually case here? There also is difference between using weak algorithms and weakening them. Former is taking place in public, latter is not...

(*) You really shouldn't, but there is no way of knowing. That's why I earlier said [in #4791] that you, as responsible peer in communication, effectively take an *obligation* to live up to declared security promise.

**Comment by richsalz** [comment]
Secure by default is being responsible; I can live with that. :)

We provide hooks to override all other policies.  Why not this one?

**Comment by dot-asm** [comment]
> We provide hooks to override all other policies. Why not this one?

Once again, there is difference between permitting weak algorithms and *weakening* ones, specifically otherwise strong ones. Because former takes place in public [through negotiation, so that peer has every change to choose to not engage in communication if deemed inappropriate], while latter is done out of peer's sight, it has to *trust* the other party. One again, it is and should be viewed as *obligation*, because of this asymmetry. Besides, do we actually aim for just digging holes deeper without challenging it? I mean if there is another hook that would put aside publicly presented security parameter, then maybe we should rather get rid of it than refer to it as "best practice". BTW, following suggested why-not-this-one logic it was inappropriate for example to remove option for shared DH keys.

**Comment by t-j-h** [comment]
Your "conceptual difference" is not relevant IMHO. We don't provide any security guarantee on the key itself or how it is used, stored, shared/not-shared, nor do we provide a list of trusted CAs, nor a lot of other things which impact operational security. We provide a tool and provide a whole range of options which allow it to be used in a range of different ways. When the original multi-prime PR went in place you argued this one - and that PR was approved and accepted - and then you added the limits in a separate PR which was rushed it without wider discussion. This PR restores the capability of the user being able to make that choice - and it is up to the user to do that.

**Comment by dot-asm** [comment]
> We don't provide any security guarantee on the key itself

Then why do we do primality checks? If we don't actually leave *any* security guarantee then we might as well omit them. [Or provide option to, as that's what effectively suggested.] It would work in sense that you'll be able to communicate with such server or generate verifiable signatures, and unsuspecting user wouldn't actually know. While we indeed do not make any assumptions about how key material is stored (or maintain certificate store content), we do undertake the responsibility for getting math right, the thing that defines RSA/PKI security. And it's highly relevant.

> When the original multi-prime PR went in place you argued this one - and that PR was approved and accepted - and then you added the limits in a separate PR which was rushed it without wider discussion.

"rushed it without wider discussion"? Get your facts straight! Things I'm saying here is effectively repetition of that was already said in #4241 in course of *days*, and even over and over. Well, in different words, but the essence is all the same, that excessive amount of primes is bad idea. *And* several team member participants did *not* dispute the suggestion that leaving it uncapped is a valid concern, rather contrary. [I wish there were more members engaged in discussion, but I'm not in position to drag people by ears.] Also note that #4241 was accepted with comment "we are in disagreement about this" that I insisted on. Obviously with intention to submit request to discuss it separately, so how come #4791 is a surprise? As for "rushing" it, I left it unmerged past moment of approval for 2 days, enough time to engage in discussion.

As already said, I yearn to hear more opinions on this, so let's hear it! Should we or should we not facilitate/participate/endorse algorithm *weakening* that is not transparent to peer? [Just in case, I'm unclear about @richsalz's position, because last remark ended with question and it's not clear if my answer is viewed as meaningful/relevant.]

**Comment by InfoHunter** [comment]
> Then why do we do primality checks?

That is another story which is totally different. Primality is a MUST for doing RSA algorithm right, for instance, as what RFC 8017 describes. But for multi-prime RSA, RFC 8077 didn't say a 2048-bit key MUST NOT have more than 3 primes. And giving a 2048-bit key with more than 3 primes does not disobey anything, in fact it's very legal to do so. This is very different from something like SSLv2, because there's almost no one in the world requires SSLv2 and SSLv2 is explicitly prohibited by RFC 6176 which is very different from multi-prime RSA: for multi-prime RSA, the requirements for more-than-3-prime-2048-bit keys are very very huge. And you may think these requirements are indefensible, but they are not and the users who want this feature are smart enough to figure out how to mitigate the risk of using a 10-prime-2048-bit key (of course by other non-technical method but not purely math...).

To define a thing is secure or not, you can't just conclude the results by *math* only. A 10-prime-2048-bit key is much easier to break than a 3-prime-2048-bit key from the point of math (everyone knows this), but it doesn't mean a 10-prime-2048-bit key is less secure than a 3-prime-2048-bit key from the point of how it's being used practically.

**Comment by p-steuer** [comment]
Just a contribution to the ongoing discussion on whether users should be able to use weak crypto i.e. not necessarily relevant to this PR:

Im in favour of enforcing a *minimum* level of security (80, 112 bits?). Algorithms (or paramaters) that do not offer this level of security according to the current state of knowledge should be disabled at compile-time.

- This allows the openssl default configuration to run in a production environment: It is guaranteed that said algorithms/paramaters will never be used, neither by accident nor by intention. So even if the minimum security level is inadvertently/intentionally used, it should not lead to a *catastrophic* failure.

- This allows users who want to test things and play around e.g. for educational purpose to run a custom openssl build.

Question is whether minimum security level should be enforced by TLS or already at libcrypto level. Id vote for the latter: as soon as the code becomes part of the object there might be ways to execute it.

**Comment by dot-asm** [comment]
> > Then why do we do primality checks?

> That is another story which is totally different.

What I'm saying that that primality of each factor is what gives RSA its strength. And it's same story all along.

> Primality is a MUST for doing RSA algorithm right...

Exactly. I have feeling that obvious has to be stated, that it's a must not because you can't perform RSA calculations if p and q are not prime, but because if they are not, then key doesn't provide claimed security qualities. [Well, there is ambiguity in this statement of following character. It's not exactly just working with *arbitrary* non-prime p and q, it's not as trivial(*), but still, their primality is not actually an *absolute* requirement.]

> To define a thing is secure or not, you can't just conclude the results by math only.

Correct, math is not sufficient component. But it's *absolutely* necessary one! Or in other words, broken math does mean broken security, *no matter* what one otherwise does [or chooses to believe]. Any attempt to put math aside can not be defended.

(*) From this angle, originating question, "why do we do primality checks", can be viewed as somewhat misleading.

**Comment by InfoHunter** [comment]
> Or in other words, broken math does mean broken security, no matter what one otherwise does

The word 'security' in your context is 'absolute security', but sometimes I just need 'partial security' and that 'partial' which the security I lost gives me the promotion of performance. And I prefer the performance than the full security in this scenario.

That is to say:

- If I use a 3-prime-2048 key to encrypt a plain text, and the cipher text will be safe. But the performance will only raise by 25%.
- If I use a 10-prime-2048 key to encrypt a plain text, the cipher text *may* eventually break by some one and the plain text *may* leaks. But the performance will raise by around 300%.

For some cases, if the plain text is not that important and the performance is critical, I will choose the 300% option, because that saves money on hardware investment and there is no evidence to show the 10-prime-2048 must be break by someone, for instance no one is interested to break this key. And you may argue if the plain text is not so important why should it be encrypted at all? That's complicated and most of the answers are not related to techniques, for example there might be some laws forcing this or some standards forbidding to use unencrypted traffic...

This is why I said 10-prime-2048 might not be 'insecure' than a 3-prime-2048 key if the user doesn't care about the 'absolute' security but only needs a partial security. But PR #4791 deprive the users of the rights of generating such a 10-prime-2048 key, as what @t-j-h said in this PR, it's up to the users to decide what to do.

**Comment by InfoHunter** [comment]
And I just don't understand, will it damage something if OpenSSL provides a non-default option to let the users generate weak multi-prime keys? Just like MD5/SHA1 is not secure but it's still widely used if you just want to verify a file's integrity.

**Comment by InfoHunter** [comment]
@paulidale , the docs are improved, not sure if it's the proper place for the explanations of why more prime makes the key easy to break...

**Comment by dot-asm** [comment]
> @p-steuer: This allows users who want to test things and play around e.g. for educational purpose to run a custom openssl build.

I for one am not convinced that OpenSSL, a project recognized as *infrastructure-critical*[!], has to provide even educational platform. But even if we assume that, if you think about it, you'll have to recognize and acknowledge that it's only educational if student can actually experience PKI weakness. And this implies that student should be able to perform the exercise on commodity hardware. Which in turn means prime sizes in 32-64 bits interval. [Which by the way goes against #4547.] Well, in the context of multi-prime one can probably argue that it might be acceptable to permit arbitrary slicing for keys that are considered to be unsafe to start with, e.g. 512-bit key sliced to up to 16 primes. Rationale would be that peer wouldn't get any illusions about key being insecure.

**Comment by dot-asm** [comment]
> Just like MD5/SHA1 is not secure but it's still widely used if you just want to verify a file's integrity. [How is it different?]

Difference is right in the question, "algorithm weakening that is *not transparent* to peer". In cases like MD5/SHA1 peer is brutally aware of the symmetric algorithm choice, not so much about private key properties. Ask yourself, why RSA2048 and not RSA1024 [or RSA512]? Because *consensus* is that RSA1024 [let alone RSA512] doesn't provide *adequate* security by today standards. And that's two primes, no buts! If you don't consider that consensual opinion matters to you, then just don't use RSA2048, stick with RSA1024 [or RSA512]. Because *otherwise*, i.e. by excessively slicing 2048-bit key, you are *deceiving* peers.

**Comment by t-j-h** [comment]
The security of the private key is not transparent to the peer - and it is absolutely criticial to the overall security of the connection. No guarantee about that is provided (nor should it be).

This is an option that is not on by default that allows the server user to make a choice.

There has never ever been a guarantee in a TLS connection with use of RSA that there will only be two primes. Never. That is entirely up to the peer (holding the private key) to decide.

**Comment by richsalz** [comment]
There is also no guarantee that the server's key hasn't been given to government agencies.

**Comment by p-steuer** [comment]
>> @p-steuer: This allows users who want to test things and play around e.g. for educational purpose to run a custom openssl build.

> I for one am not convinced that OpenSSL, a project recognized as infrastructure-critical[!], has to provide even educational platform.

i didnt intend to say it should. My point was rather that the needs of users who use openssl in production should take precedence over the needs of users who use openssl for whatever reason.

That said, i dont have objections against multi prime rsa in general but wont comment on parameter choices.

**Comment by dot-asm** [comment]
> There is also no guarantee that the server's key hasn't been given to government agencies.

Could you clarify? Does it mean that you suggest that since *we* can't leave such guarantee, we don't have to be diligent?

**Comment by richsalz** [comment]
I claim we're being diligent by making the default  "secure" and that's enough.

**Comment by p-steuer** [comment]
I dont get the no guarantee thing. I mean when you make cars, you can not guarantee that future owner will not manipulate it. But to install a switch that increases the performance for the price of higher than advertised emission values, is another story.

In other words, the question is whether it is okay for libcrypto to offer such functionality and your answer will be a precedent for future submissions.

**Comment by richsalz** [comment]
I see @dot-asm 's point about this being invisibly weaker. But by default we do the right thing, and we allow applications to install their own config to use more/weaker primes. Someone who wants to do this, can just comment-out the checks anyway. Just like we're not providing a default that lets weak thing happen.

File: doc/man1/genrsa.pod [hunk 2]
```diff
@@ -112,6 +121,15 @@ Because key generation is a random process the time taken to generate a key
 may vary somewhat. But in general, more primes lead to less generation time
 of a key.
 
+For multi-prime RSA, which is defined in RFC 8017, unlike normal 2-prime RSA,
+it's security depends on both how many primes are used and how long the RSA
+modulus is. But it's still estimated by the difficulty of factoring the modulus.
+In simple terms, the more primes a multi-prime RSA key has, the easier for it
+being factored. Here is a link describing the security of multi-prime RSA
+generally: https://www.imperialviolet.org/2011/04/09/multiprime.html. And if
+you want to understand the very details, here is the right place to go to:
```
Comment (paulidale) [inline doc/man1/genrsa.pod]: s/you want to understand the very details, here is the right place to go to/you want to know more, this link provides the requisite details/

File: doc/man1/rsa.pod [hunk 3]
```diff
@@ -117,6 +118,10 @@ This option prints out the value of the modulus of the key.
 
 This option checks the consistency of an RSA private key.
 
+=item B<-insecure>
+
+This option is used for checking insecure multi-prime RSA private keys.
```
Comment (paulidale) [inline doc/man1/rsa.pod]: s/checking/allowing/ or s/checking/permitting/

File: doc/man1/rsa.pod [hunk 4]
```diff
@@ -195,6 +200,10 @@ Output the public part of a private key in B<RSAPublicKey> format:
 
  openssl rsa -in key.pem -RSAPublicKey_out -out pubkey.pem
 
+To check an insecure multi-prime private which has 8 primes with 2048-bit:
```
Comment (paulidale) [inline doc/man1/rsa.pod]: s/check/allow/ or s/check/permit/

File: doc/man3/RSA_generate_key.pod [hunk 5]
```diff
@@ -42,6 +42,9 @@ B<primes> depends on modulus bit length:
    ------+--------+--------+-------
      2   |   3    |   4    |   5
 
+A flag B<RSA_FLAG_INSECURE_PRIMES> can be used to switch off the above policy
```
Comment (paulidale) [inline doc/man3/RSA_generate_key.pod]: s/A flag/The flag/

File: doc/man1/genpkey.pod [hunk 6]
```diff
@@ -109,6 +109,11 @@ The number of bits in the generated key. If not specified 1024 is used.
 
 The number of primes in the generated key. If not specified 2 is used.
 
+=item B<rsa_keygen_mode:mode>
+
+The mode in use when generating a key, either 'insecure' or 'secure' is
+availiable at current stage. If not specified 'secure' is used.
```
Comment (paulidale) [inline doc/man1/genpkey.pod]: s/at current/at the current/

File: doc/man1/genrsa.pod [hunk 7]
```diff
@@ -96,6 +97,14 @@ RSA key, which is defined in RFC 8017.
 The size of the private key to generate in bits. This must be the last option
 specified. The default is 2048.
 
+=item B<-insecure>
+
+By default OpenSSL will force users to generate secure multi-prime keys only.
+That is to say, the number of primes is limited for a given length of modulus
+based on a security policy, see L<RSA_generate_multi_prime_key(3)> for details
+of that security policy. But sometimes users may need to generate more primes
+for a private key, this option can turn off the default security policy.
```
Comment (paulidale) [inline doc/man1/genrsa.pod]: s/can turn off/turns off/ or s/can turn off/disables/

File: doc/man1/genrsa.pod [hunk 8]
```diff
@@ -112,6 +121,15 @@ Because key generation is a random process the time taken to generate a key
 may vary somewhat. But in general, more primes lead to less generation time
 of a key.
 
+For multi-prime RSA, which is defined in RFC 8017, unlike normal 2-prime RSA,
+it's security depends on both how many primes are used and how long the RSA
+modulus is. But it's still estimated by the difficulty of factoring the modulus.
+In simple terms, the more primes a multi-prime RSA key has, the easier for it
```
Comment (paulidale) [inline doc/man1/genrsa.pod]: s/easier for it/easier it is for it/

**Comment by paulidale** [comment]
I probably could have laid those suggestions out better :(

**Comment by InfoHunter** [comment]
All doc issues have been addressed, @paulidale

**Comment by mattcaswell** [comment]
Do we need an OMC vote to resolve this disagreement?

**Comment by t-j-h** [comment]
No - there is no -1 vote from an OMC member so no requirement for an OMC vote to be held - and there are two approvals on record so I expect that @richsalz will be pushing this into git shortly.

**Comment by dot-asm** [comment]
I would still like to here more opinions. To recap, the question is should we or should we not facilitate/participate/endorse algorithm weakening that is not transparent to peer?

**Comment by dot-asm** [comment]
> @richsalz: Someone who wants to do this, can just comment-out the checks anyway.

I don't follow. Or rather I find this utterly contradictory. If someone can uncomment something *anyway*, then why *this* has to go into repository? Or is  suggestion actually is that if somebody doesn't like to have this option in production build, then they should uncomment something?

**Comment by dot-asm** [comment]
> I disagree with the absolutist "no" position

Do you subscribe to absolutist "yes" position?

**Comment by richsalz** [comment]
Secure by default is my view.

**Comment by richsalz** [comment]
We have strenuous objections from a long-term member of the community. I know we *can* merge this, but I could like to see if we can reach accomodation. My guess is no. But I have no plans to merge this right now; someone else, of course, can do so.

**Comment by t-j-h** [comment]
My concern is that we discussed this at length on the original PR #4241 and we basically agreed to disagree. 

And then another PR #4791 went it rapidly and without the same length of discussion - and it also had clearly stated objections - objections that were essentially ignored once there was an approval - despite the fact that the discussion continuing and that status being clear. 

And now we have the same discussion from the original PR being repeated here. This is a fundamental disagreement - and the views have been clearly expressed. Those strenuous objections on the original PR were discussed. What Andy is seeking here is to basically re-argue the original PR and to have this PR treated differently to the PR that supported his view (#4791) - which went in despite clearly noted objections.

This isn't a matter of rushing things - that argument was expressed at length - and nothing new has been added to that discussion in this PR, nor have the positions expressed by those who have commented changed in any way.

**Comment by paulidale** [comment]
A thought: would a warning to the effect that "this OpenSSL build includes insecure options _caveat emptor_" be appropriate?

I'm thinking after the initial `config` command and possibly from the command line tools as well.  In this case, the `rsa` and `genrsa` commands and possibly others.  This might help to partly address the _silent weakness_ concerns.

File: doc/man1/genrsa.pod [hunk 9]
```diff
@@ -112,6 +121,15 @@ Because key generation is a random process the time taken to generate a key
 may vary somewhat. But in general, more primes lead to less generation time
 of a key.
 
+For multi-prime RSA, which is defined in RFC 8017, unlike normal 2-prime RSA,
+its security depends on both how many primes are used and how long the RSA
+modulus is. But it's still estimated by the difficulty of factoring the modulus.
+In simple terms, the more primes a multi-prime RSA key has, the easier it is for
```
Comment (kaduk) [inline doc/man1/genrsa.pod]: I don't like starting a sentence with "But" and no comma; maybe "The security is still related to the difficulty of factoring the modulus -- in simple terms, [...]"

**Comment by InfoHunter** [comment]
Address `However,` and the `But...` sentence

**Comment by InfoHunter** [comment]
> A thought: would a warning to the effect that "this OpenSSL build includes insecure options caveat emptor" be appropriate?

If doing so, it should be done for all insecure options, like SSLv3 or RC4...

**Comment by t-j-h** [comment]
@kaduk - a warning such as that serves no purpose unless it is visible to the people that would care to see such an option - and that is not something an SDK can do - we return errors to applications - and it is up to the application as to how it is handled.

It also doesn't address the concern that @dot-asm raised - as the client connecting has no way of knowing that the server has been configured this way - so even if the server issued a warning - it would be warning the wrong party in the interaction.

I fully appreciate @dot-asm viewpoint - I just disagree with a base assumption/premise in terms of what OpenSSL is meant to be doing - and I also think that picking on this particular feature while ignoring the many other much easier insecure configuration options is both misfocused and also not something that is part of the responsibility of OpenSSL.

**Comment by ekasper** [comment]
> To recap, the question is should we or should we not facilitate/participate/endorse algorithm weakening that is not transparent to peer?

I'm with @dot-asm on this one. -1, we should not, let's vote.

First, multi-prime RSA is an advanced performance optimization feature for users that are extremely sensitive to performance. 20-25% is not insignificant when you are in this position. The PR makes the claim that "a lot of users" will be unable to use this feature - this looks unsubstantiated to me.

Second, you simply cannot have RSA for free. Adding more primes is the wrong trade-off. If you want to trade performance for security, slide the key length, thus making your security choice transparent. If you have performance problems with longer keys, you may want to evaluate against EC(IES/DSA) instead of RSA keys with more primes. Pretending to have 2048-bit equivalent security when you don't is misleading, and dangerous to the users who _cannot tell that an insecure configuration has been chosen_. In that sense, it's very different from "I went through extra hoops to enable DES". Would you want your banking provider to silently compose their SSL certificate private key of small primes? 

Third, since it is a performance optimization feature, the temptation to abuse it will be high if no obstacles are put in place. I do not consider a flag an obstacle - we have plenty of evidence that "if it compiles it will be (mis)used". This'll be a highly dangerous feature if left unrestricted, and I'd only be left looking forward to the high profile IEEE S&P / ACM CCS / Usenix paper that scavenges the internet for weak multi-prime keys. We can't stop people from patching/writing their own weak(ened) key generators but I do not think we should endorse this with legal APIs.

Whatever is decided, OpenSSL has _way_ too many compile-time knobs already and so I am very much against adding more, especially for this purpose.

**Comment by vdukhovni** [comment]
-1. (Agreeing with Emilia and Andy)

I should mention that the main issue is not about generation of weaker multi-prime keys, but rather support for the (CRT-optimized) use of weaker multi-prime keys.  If someone wants to generate a private key that OpenSSL will not be able to use, they can do that with some other tool and just deploy the key. The real question is whether we allow the use of said multi-prime keys when the individual prime sizes weaken the RSA modulus.

Now perhaps the (e.g.) 2048-bit key is required to comply with an overzealous standard that demands needlessly strong security theatre, or because the device needs to support clients (browsers?) that are unwilling to use clearly weak keys, but might be "fooled" by apparently 2048-bit keys that under covers trade-off security for performance.  OK, perhaps that may reasonable in some cases, but it is not clear that OpenSSL should support that out of the box without code modifications. Indeed it is far from clear that OpenSSL is actually a good choice of library for resource-constrained embedded applications. I would expect that other toolkits are more appropriate in that space.  OpenSSL's market is I think primarily general purpose computers and high-performance embedded applications (firewalls, proxies, ...). I would not be worried about ceding the resource-constrained space to more specialized light-weight libraries.

**Comment by levitte** [comment]
> Would you want your banking provider to silently compose their SSL certificate private key of small primes?

No...  but surely, our limiting the *production* of such keys doesn't really make a difference there, does it?  However, I cannot see anywhere that we allow *receiving* such insecure keys, especially on the wire.  I can only assume that other tools with similar functionality do their own security checks...

After all, insecure keys can as well be produced with another tool, and then attempted to be used with OpenSSL.  Will they work or not as input?  Isn't that the interesting consideration to make?

**Comment by bernd-edlinger** [comment]
Yes, I think we should at least have a simple check on imported private keys.
For instance lowering RSA_MAX_PRIME_NUM from 16 to 5 would be reasonable.
And checking that, in the moment that check is needed, seems to be nowhere...

**Comment by ekasper** [comment]
Oh good point, yes we shouldn't allow these private keys to be used either. However, levitte@, note that one would normally _receive_  a public key on the wire, and that is indistinguishable from a normal 2-prime key - and therefore if one manages to e.g. patch together a weak multiprime serverside implementation, then the client can do nothing to protect itself. While we can't stop other implementations from abusing this, we should not contribute to this behaviour either. The appropriate mechanism for performance/security trade offs already exists and is called key size.

**Comment by bernd-edlinger** [comment]
See #4862 the check there should prevent any use of a bogus private key in private key ops
and a possible buffer overrun at the same time.
Still to do: change the rather high 16 prime limit to a more sensible number.

**Comment by InfoHunter** [comment]
I still can't understand how could it harm the very *holy* OpenSSL project by adding an `-insecure` option or flag to this toolkit (for multi-prime RSA feature of course). I don't mean it should be *encouraged* or *endorsed* to use weak multi-prime RSA key, I never said or expressed that. What I am doing is only trying to provide a choice for the users to fulfill their requirements if the user knows what he/she is doing. And this provide this product versatility. (And you also need to realize that the word *users* here not only referring to the users you are familiar with...)

Again why this PR is here, as @t-j-h mentioned above:

```
And then another PR #4791 went it rapidly and without the same length of discussion - and it also had clearly stated objections - objections that were essentially ignored once there was an approval - despite the fact that the discussion continuing and that status being clear.
```

The only difference I see is #4241 and this PR are opened by a non-team-member who can't do the rushed merge as what #4791 did. (This PR has enough approvals but still goes into a vote)

Anyway, I am enough with this unfair game.

**Comment by InfoHunter** [comment]
> Oh good point, yes we shouldn't allow these private keys to be used either.

> 20-25% is not insignificant when you are in this position. The PR makes the claim that "a lot of users" will be unable to use this feature - this looks unsubstantiated to me.

I suggest you to talk more with other OMC members who know better about the markets and the users outside western developed countries - and you will get different knowledges and will not make these assertions so subjectively - we have different scenarios or problems to face...

**Comment by dot-asm** [comment]
Additional note about not providing "any security guarantee". Just in case[!] remark was reference to the fact that primality test is probabilistic. While it's true, it's also true that we are committed to doing everything that is *consensually* considered necessary to achieve acceptable level. But "consentually" in last sentence doesn't refer to consensus within team, but within expert community.

**Comment by ekasper** [comment]
To close the loop here: the OMC voted against accepting this PR.

I believe a logical consequence would be to also place proper limits on the use of weakened keys. @bernd-edlinger, you mention a "still to do" - do you plan to do that?

**Comment by bernd-edlinger** [comment]
I did #4862 and #4905 after my "still to do" comment: Both together limit the number of primes on
any imported key to 5 which is what considered the absolute minimum we have to do.

**Comment by ekasper** [comment]
Ah, I'd missed #4905. Great, thanks!

**Pull request closed without merging**